Dockerfile.ci

```diff
@@ -309,6 +322,8 @@ RUN if [[ -n "${ADDITIONAL_PYTHON_DEPS}" ]]; then \
         pip install ${ADDITIONAL_PYTHON_DEPS}; \
     fi
 
+RUN pip install pyodbc
```

Comment:
hmm isn't this already in setup.py? Why do it separately?

Reply:
thanks for the tips over slack :) - moved it in https://github.com/apache/airflow/pull/9973/commits/f2cb6ccde0c8902c02e1cedee4b225ed93ed03d0

Reply:
should I increment `PIP_DEPENDENCIES_EPOCH_NUMBER` ?

Reply:
oh yea, incremenet it so that it bursts the cache

Reply:
Yoy only need to do it if you remove some dependencies. For the added ones it's not needed (and even harmful - slower builds precisely because cache is dropped)..

Reply:
aah I didn't know that. How does Docker know something has changed, do we store the setup.py SHA in dockerfile?

Reply:
Docker does it on its own - whenever you add anything from local context with COPY or ADD it calculates hashes of the added files and will invalidate the cache when they're different from the ones used to build it previously.

That's why we are adding setup.py in Dockerfile separately at some point in Dockerfile - to only invalidate the cache if setup.py changes (but I also do a smart pre-caching where I install all deps from github first  - so that changing setup.py will not invalidate all installed pip reps.

This results with much faster rebuilds when we add or upgrade something in setup.py (this way only the changed deps got added/upgraded) but when we remove something, we need to increase the EPOCH, so that when it's merged to master, it will reinstall it from the scratch.

Reply:
cool, thank you

Reply:
aah I didn't see that we were COPYing setup.py separately, good work in that case :-D . I know about Docker handling hash itself when using COPY or ADD.